astropy/io/fits/util.py

```diff
@@ -982,3 +984,74 @@ def _get_array_mmap(array):
         if isinstance(base.base, mmap.mmap):
             return base.base
         base = base.base
+
+
+def get_posargs(function):
+    """Helper function to retrieve the positional parameters from a
+    function
+
+    inspect.getargspec is deprecated in Python 3, while
+    inspect.signature does not exist in Python 2. This function caters
+    for both.
+
+    """
+    try:
+        posargs = list(inspect.signature(function).parameters.keys())
+    except AttributeError:
+        posargs = inspect.getargspec(function).args
+
+    return posargs
+
+
+def replaced_argument(oldkeyword, newkeyword):
+    """Decorate a function to replace 'oldkeyword' by 'newkeyword, but
+    allow for (deprecated) use of 'oldkeyword''
+
+    The decorated function should follow the new signature: the
+    definition should use the newkeyword instead of the oldkeyword.
+
+    The decorator
+
+    - accepts oldkeyword behind the scenes. If given instead of
+      newkeyword, an AstropyDeprecationWarning is given, and the
+      function is called with the value of oldkeyword.
+
+    - verifies that oldkeyword and newkeyword are not used at the same
+      time: a TypeError will be raised. This is the case even when the
+      value is passed as a positional parameter.
+
+    """
+    def decorator(function):
+        @functools.wraps(function)
+        def wrapper(*args, **kwargs):
+            # Determine the position of the possible newkeyword
+            # positional parameter
+            posargs = get_posargs(function)
```

Comment:
You could do this in a more lazy fashion. I don't know how fast/slow ``inspect`` is but it might be faster to do something like this:

```
def decorator(function):
    posarg = None
    @functools.wraps(function)
    def wrapper(*args, **kwargs):
        nonlocal posarg
        if posarg is None:
            posarg = random.randint(0, 100) # put your logic here
        ...
    return wrapper
```